stellargraph/mapper/sequences.py

```diff
@@ -527,3 +530,128 @@ def __len__(self):
+    Use this class with the Keras methods :meth:`keras.Model.fit`,
+        :meth:`keras.Model.evaluate`, and
+        :meth:`keras.Model.predict`,
+
+    This class should be created using the `.flow(...)` method of
+    :class:`GraphGenerator`.
+
+    Args:
+        graphs (list)): The graphs as StellarGraph objects.
+        targets (np.ndarray, optional): An optional array of graph targets of size (N x C),
+            where N is the number of graphs and C is the target size (e.g., number of classes.)
+        normalize_adj (bool, optional): Specifies whether the adjacency matrix for each graph should
+            be normalized or not. The default is True.
+        batch_size (int, optional): The batch size. It defaults to 1.
+        name (str, optional): An optional name for this generator object.
+    """
+
+    def __init__(
+        self, graphs, targets=None, normalize_adj=True, batch_size=1, name=None
+    ):
+
+        self.name = name
+        self.graphs = graphs
+        self.normalize_adj = normalize_adj
+        self.targets = targets
+        self.batch_size = batch_size
+        # we assume that all graphs have node features of the same dimensionality
+        self.node_features_size = graphs[0].node_features(graphs[0].nodes()).shape[1]
+
+        if targets is not None:
+            if len(graphs) != len(targets):
+                raise ValueError(
+                    "When passed together targets and the number of graphs should be the same length."
+                )
+
+            self.targets = np.asanyarray(targets)
+
+        self.on_epoch_end()
+
+    def __len__(self):
+        return int(np.ceil(len(self.graphs) / self.batch_size))
+
+    def __normalize_adj(self, adj):
+        adj = adj.tolil()
+        adj.setdiag(1)  # add self loops
+        degree_matrix_diag = 1.0 / np.sqrt(adj.sum(axis=1))
+        degree_matrix_diag = np.squeeze(np.asarray(degree_matrix_diag))
+        degree_matrix = sparse.lil_matrix(adj.shape)
+        degree_matrix.setdiag(degree_matrix_diag)
+        degree_matrix = degree_matrix.tocsr()
+        adj = degree_matrix @ adj @ degree_matrix
+        return adj
+
+    def __getitem__(self, index):
+        graphs = self.graphs[
+            index * self.batch_size : (index * self.batch_size) + self.batch_size
+        ]
+        adj_graphs = [graph.to_adjacency_matrix() for graph in graphs]
+
+        # The number of nodes for the largest graph in the batch. We are going to pad with 0 rows and columns
+        # the adjacency and node feature matrices (only the rows in this case) to equal in size the adjacency and
+        # feature matrices of the largest graph.
+        max_nodes = max([graph.number_of_nodes() for graph in graphs])
+
+        # The operations to normalize the adjacency matrix are a bit slow.
+        # FIXME: Either optimize this or implement as a layer.
+        if self.normalize_adj:
+            adj_graphs = [self.__normalize_adj(adj).toarray() for adj in adj_graphs]
```

Comment:
I think we can do this densification better: at the the moment we start with sparse (normalised) adjacency matrices, convert these to dense, and then fill out the rest of the each one with zeros to create new dense matrices. However, sparse matrices are really good at representing zeros cheaply, and so adding extra zero rows and columns can be done very efficiently by just playing with their shapes.

```python
def _resize_csr_square(m, new_dim):
    return sparse.csr_matrix((m.data, m.indices, m.indptr), shape=(new_dim, new_dim))

    ...

            adj_graphs = [self.__normalize_adj(adj) for adj in adj_graphs]

    ...

        adj_graph = np.stack([_resize_csr_square(adj, max_nodes).toarray() for adj in adj_graphs])
```

Reply:
I didn't know this. I'll update as suggested.

Reply:
Resizing the csr matrix as suggested does not work. The following program,

```
import scipy.sparse as sps
a = sps.rand(10, 10, density=0.01, format='csr')
sps.csr_matrix((a.data, a.indices, a.indptr), shape=(12,12))

```
raises the following exception,

`ValueError: index pointer size (11) should be (13)`

I'm using scipy version 1.4.0. Something must have changed in scipy that this no longer works as I see this has been suggested as a solution at https://stackoverflow.com/questions/26462617/adding-a-column-of-zeroes-to-a-csr-matrix as well.

However, calling `a.resize((12,12))` seems to work as expected so I might give this one a try.

Reply:
Ah, sorry for the misleading advice; that answer is indeed what I was working off. I looked for a resizing function like that but didn't notice it; oops. Looking at the implementation it does seem like it's a moderately complicated operation for `csr` matrices, so definitely good to use that 👍